arangod/Aql/Expression.cpp

```diff
@@ -358,24 +358,24 @@ void Expression::initAccessor(ExpressionContext& ctx) {
 }
 
 /// @brief prepare the expression for execution
-void Expression::prepareForExecution(ExpressionContext& ctx) {
+void Expression::prepareForExecution() {
   TRI_ASSERT(_type != UNPROCESSED);
 
   switch (_type) {
     case JSON: {
       if (_data == nullptr) {
         // generate a constant value
-        transaction::BuilderLeaser builder(&ctx.trx());
-        _node->toVelocyPackValue(*builder.get());
+        velocypack::Builder builder;
+        _node->toVelocyPackValue(builder);
 
-        _data = new uint8_t[static_cast<size_t>(builder->size())];
-        memcpy(_data, builder->data(), static_cast<size_t>(builder->size()));
+        _data = new uint8_t[static_cast<size_t>(builder.size())];
+        memcpy(_data, builder.data(), static_cast<size_t>(builder.size()));
```

Comment:
I made the method "prepareForExecution" public and removed the "trx" parameter from its signature.
I wanted to avoid that this function would store any data permanently inside something owned by the transaction, because the expressions now can be used across multiple transactions. That's why no BuilderLeaser is used here anymore, as it depends on the transaction's context.
I tried to reduce the overhead of the function a bit after your comment, by at least avoiding the shared_ptr reference count increase when creating a new Builder. Now it is creating a Builder from an existing Buffer, which is slightly more efficient.

Reply:
How about we have two `prepareForExecution` overloads, one that takes a builder and one that doesn't. The one without arguments can create a local builder (just like you do right now) and call the second overload passing that builder. However, `execute` could than pass a builder from the context's BuilderLeaser.